bonsai/models/qwen3/tests/run_model.py

```diff
@@ -0,0 +1,66 @@
+import math
+import os
+import time
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from flax import nnx
+from huggingface_hub import snapshot_download
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from bonsai.models.qwen3 import modeling, params
+
+model_name = "Qwen/Qwen3-0.6B"
+MODEL_CP_PATH = "/tmp/models-bonsai/" + model_name.split("/")[1]
+if not os.path.isdir(MODEL_CP_PATH):
+    snapshot_download(model_name, local_dir=MODEL_CP_PATH)
+
+
+def tokenize(tokenizer, input: list[str]):
+    pad_idx = tokenizer.pad_token_id
+    lines = [
+        tokenizer.apply_chat_template(
+            [{"role": "user", "content": line}], tokenize=False, add_generation_prompt=True, enable_thinking=False
+        )
+        for line in input
+    ]
+    lines = [tokenizer.encode(line, add_special_tokens=False) for line in lines]
+    max_len = 2 ** math.ceil(math.log2(max(max(len(line) for line in lines), 1)))  # smallest next power of 2.
+    output = [[pad_idx] * (max_len - len(line)) + line for line in lines]
+    return np.array(output)
+
+
+query = ["Why is the sky blue instead of any other color like purple?", "Who is Harry Potter?"]
+tokenizer = AutoTokenizer.from_pretrained(MODEL_CP_PATH)
+tokens = jnp.array(tokenize(tokenizer, query))
+batch_size, token_len = tokens.shape
+
+config = modeling.ModelConfig.qwen3_0_6b()  # pick correponding config based on model version
+model = params.create_model_from_safe_tensors(MODEL_CP_PATH, config)
+cache = modeling.init_cache(num_layers=28, batch_size=batch_size, cache_size=256, num_kv_heads=8, head_dim=128)
+graphdef, state = nnx.split((model, cache))
+# Better performance from flattened jax state due to no pytree trasversals.
+state = jax.tree.leaves(state)
+
+# prefill
+next_tokens, state = modeling.forward(graphdef, state, tokens, True)
+
+# decode
+tokens_list = [next_tokens]
+# Run `xprof --logdir /tmp/profile-data/` to see the program traces.
+for i in range(11):
+    if i == 1:  # Avoid XLA warmup time when measuring trace.
+        jax.profiler.start_trace("/tmp/profile-data")
+    next_tokens, state = modeling.forward(graphdef, state, next_tokens, False, token_len + i)
+    tokens_list.append(next_tokens)
+    if i == 5:  # profile steps 1-5
+        jax.block_until_ready(tokens_list)
+        jax.profiler.stop_trace()
+        t = time.perf_counter()  # measure steps 6-10
+jax.block_until_ready(tokens_list)
+print(f"{time.perf_counter() - t:.4f} s")
+tokens_list = jnp.squeeze(jnp.array(tokens_list), axis=2).T
+for i, q in enumerate(query):
+    print(f"User:\n {q}")
+    print(f"Answer:\n {''.join([tokenizer.decode(t) for t in tokens_list[i]])}\n\n")
```

Comment:
nit: tokens_list.ndim == 2, so ideally two indices to slice into it `tokens_list[i, :]`